### DISCUSSION TITLE
Entity discovery on TsMorphMetadataProvider broken since 6.2.2

### DISCUSSION THREAD

**Discussion by sebastiaanviaene** [mikro-orm/mikro-orm#5536] [General]
### Describe the bug

I have the following orm config for entity discovery. Since 6.2.2 and specifically due to [this commit](https://github.com/mikro-orm/mikro-orm/commit/d64be7ef5ba7c2a655ea014992905d8afde3c231) it now fails to discover my ts entity files

```ts
const ormConfig = defineConfig({
  entities: ['dist/**/*.entity.js'],
  entitiesTs: ['src/**/*.entity.ts'],
 ...
});

```

The error I get:

```
MetadataError: Source file './dist/src/shared/entities/batch.entity.ts' not found. Check your 'entitiesTs' option and verify you have 'compilerOptions.declaration' enabled in your 'tsconfig.json'. If you are using webpack, see https://bit.ly/35pPDNn
```

when debugging I notice it's trying to translate  `./dist/src/shared/entities/batch.entity.js` to `./dist/src/shared/entities/batch.entity.ts` but those are not in my sources array (because my source entities are not in dist ofcourse)

I found a work around but it's far from ideal. Notice the 2 different patterns due to __dirname being different for node / ts-node

```ts
const ormConfig = defineConfig({
  entities: ['**/*.entity.js', '../dist/**/*.entity.js'],
  entitiesTs: ['../../src/**/*.entity.ts', '**/*.entity.ts'],
  baseDir: path.resolve(__dirname),
  ...
  },
});
```

### Reproduction

-

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.2.2

### Node.js version

20.9.0

### Operating system

MacOS

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
I need to see a reproduction, pointing to a problematic commit won't cut it.

cc @xmajzel

---

Note that this is almost always a bad idea, `baseDir` shouldn't be dynamic (which `__dirname` is).

    baseDir: path.resolve(__dirname),

Or you say that's the workaround to get around this for you? Still, it rather points to a misconfiguration.

Also if you see paths like `/dist/src` its almost always wrong too. Or do you compile the whole `src` folder into `dist`?

**Reply by sebastiaanviaene** [reply]
Its a work around to make it work after 6.2.2

**Reply by sebastiaanviaene** [reply]
I do have src in my dist among other things

**Comment by v1d3rm3** [comment]
I believe you trying to run after compilation. I was having similar problem, i don't if it could help you. I was executing `node dist/main.js`, but i was receiving the error that you described. It have worked when i changed the command to `node --enable-source-maps dist/main.js`.

**Reply by B4nan** [reply]
The thing is that this shouldn't be needed (and I believe it wasn't needed before). The problem is most likely about the extraction of file paths from stack traces (and the matching to the TS files based on that).

If you can share a complete repro, it would help, so far we don't have anything failing (those we had are fixed).

**Comment by softimiz** [comment]
@B4nan Hello Martin,

We have the same problem on our side. We can't upgrade from 6.2.1 to 6.3.0.

Here's the result of mikro-orm debug command from 6.2.1 and 6.3.0 (entitiesTs stays the same)
```
Current MikroORM CLI configuration
 - dependencies:
   - mikro-orm 6.3.0
   - node 20.12.2
   - knex 3.1.0
   - pg 8.11.3
   - typescript 5.5.3
 - package.json found
 - ts-node enabled
 - searched config paths:
   - C:/dev/softimiz-api/src/db/mikro-orm.config.ts (found)
   - C:/dev/softimiz-api/dist/db/mikro-orm.config.js (not found)
   - C:/dev/softimiz-api/src/mikro-orm.config.ts (not found)
   - C:/dev/softimiz-api/mikro-orm.config.ts (not found)
   - C:/dev/softimiz-api/dist/mikro-orm.config.js (not found)
   - C:/dev/softimiz-api/mikro-orm.config.js (not found)
 - configuration found
 - database connection successful
 - will use `entities` array (contains 0 references and 1 paths)
   - C:/dev/softimiz-api/dist/**/*.entity.js (found)
 - could use `entitiesTs` array (contains 0 references and 1 paths)
   - C:/dev/softimiz-api/src/**/*.entity.ts (found)
```

```
Current MikroORM CLI configuration
 - dependencies:
   - mikro-orm 6.2.1
   - node 20.12.2
   - knex 3.1.0
   - pg 8.11.5
   - typescript 5.5.3
 - package.json found
 - ts-node enabled
 - searched config paths:
   - C:/dev/softimiz-api/src/db/mikro-orm.config.ts (found)
   - C:/dev/softimiz-api/dist/db/mikro-orm.config.js (not found)
   - C:/dev/softimiz-api/src/mikro-orm.config.ts (not found)
   - C:/dev/softimiz-api/mikro-orm.config.ts (not found)
   - C:/dev/softimiz-api/dist/mikro-orm.config.js (not found)
   - C:/dev/softimiz-api/mikro-orm.config.js (not found)
 - configuration found
 - database connection successful
 - will use `entities` array (contains 0 references and 1 paths)
   - C:/dev/softimiz-api/dist/**/*.entity.js (found)
 - could use `entitiesTs` array (contains 0 references and 1 paths)
   - C:/dev/softimiz-api/src/**/*.entity.ts (found)
```

I get this error starting NestJs with Mikro 6.3.0:
`Source file './dist/src/shared/crud/crud.entity.ts' not found. Check your 'entitiesTs' option and verify you have 'compilerOptions.declaration' enabled in your 'tsconfig.json'. If you are using webpack, see https://bit.ly/35pPDNn`

I went to @mikro-orm\reflection\TsMorphMetadataProvider.js and added some break points to the getSourceFile function.

The behavior with 6.2.1 is
```
tsPath = './dist/src/shared/crud/crud.entity.d.ts'
s.getFilePath() = "C:/dev/softimiz-api/src/shared/crud/crud.entity.ts"
at the first passage, source is found
```

6.3.0 Behavior:
```
tsPath = './dist/src/shared/crud/crud.entity.d.ts'
s.getFilePath() = "C:/dev/softimiz-api/src/shared/crud/crud.entity.ts"
first passage, source is undefined


tsPath = './dist/src/shared/crud/crud.entity.ts'
s.getFilePath() = "C:/dev/softimiz-api/src/shared/crud/crud.entity.ts"
second passsage, source is undefined and exception is thrown
```

Hope this helps you pinpoint the problem. Let me know if you need anything else.

**Reply by B4nan** [reply]
I need a complete reproduction, from this it's not really clear why it fails to find the source file.

**Reply by softimiz** [reply]
@B4nan 

There you go:
https://github.com/softimiz/mikro-orm-nestjs-bug-reproduction

**Reply by B4nan** [reply]
Thanks, this was really helpful. This should help: fe54d45a659c81bb2b8793085c7b005478c12a97

**Reply by softimiz** [reply]
Works like a charm with v6.3.3-dev.11 😄 Thanks Martin

**Comment by r1tsuu** [comment]
For me it works with NestJS, but only if not using SWC compiler. With SWC compiler I'm getting the same error, migrated to `ReflectMetadataProvider` and everything is ok with SWC.

**Comment by B4nan** [comment]
Would be great if those who still suffer from this problem could try latest dev version (so `v6.3.3-dev.9` or later). With fe54d45, we try to automatically strip the `outDir` from the path, so it should cover those setups where the `dist` folder is compiled inside the `src`. If not, I am yet again asking for another complete reproduction that is still not working :]

**Reply by r1tsuu** [reply]
Works for me! Thank you
